Implement multiple occlusion polygons within each TileSet occlusion layer

groud opened the pull request:
Closes https://github.com/godotengine/godot/issues/66877

This PR keeps backward compatibility, but I had to modify the API to account for more than one polygon.

I think this is mostly an enhancement, but well, the linked issue is marked as a bug, so well, I used both tags.
Not sure we still have the time for such change for 4.3, so I used the 4.x milestone.

groud committed:
Implement multiple occlusion polygons within each TileSet occlusion layer

chryan commented:
Fixed some bugs with this pull request here:
https://github.com/godotengine/godot/pull/95178

groud commented:
Rebased the PR, and fixed the issue mentioned in https://github.com/godotengine/godot/pull/95178.
Thanks for the feedback @chryan.

Mickeon commented on doc/classes/TileData.xml:
```diff
@@ -16,6 +16,13 @@
+			<return type="void" />
+			<param index="0" name="layer_id" type="int" />
+			<description>
+				Adds an occlusion polygon to the tile on the given TileSet occlusion layer.
```
It seems like doing this is consistent with the rest you wrote:
```suggestion
				Adds an occlusion polygon to the tile on the TileSet occlusion layer with index [param layer_id].
```

Mickeon commented on doc/classes/TileData.xml:
```diff
@@ -89,6 +96,25 @@
+			<param index="4" name="transpose" type="bool" />
+			<description>
+				Returns the occluder polygon at index [param polygon_index] from the TileSet occlusion layer with index [param layer_id].
+				[param flip_h], [param flip_v], and [param transpose] allow transforming the returned polygon.
```
Nitpick. It's fine even now.
```suggestion
				The [param flip_h], [param flip_v], and [param transpose] parameters can be [code]true[/code] to transform the returned polygon.
```

Mickeon commented on doc/classes/TileData.xml:
```diff
@@ -89,6 +96,25 @@
+			<return type="OccluderPolygon2D" />
+			<param index="0" name="layer_id" type="int" />
+			<param index="1" name="polygon_index" type="int" />
+			<param index="2" name="flip_h" type="bool" />
+			<param index="3" name="flip_v" type="bool" />
+			<param index="4" name="transpose" type="bool" />
```
A bit concerned about these parameters not being optional.

AThousandShips commented on scene/resources/2d/tile_set.h:
```diff
@@ -940,8 +943,17 @@ class TileData : public Object {
 	Ref<OccluderPolygon2D> get_occluder(int p_layer_id, bool p_flip_h = false, bool p_flip_v = false, bool p_transpose = false) const;
+#endif // DISABLE_DEPRECATED
+
+	void set_occluder_polygons_count(int p_layer_id, int p_polygons_count);
+	int get_occluder_polygons_count(int p_layer_id) const;
```
```suggestion
	void set_occluder_polygon_count(int p_layer_id, int p_polygons_count);
	int get_occluder_polygon_count(int p_layer_id) const;
```
English doesn't use plurals for nouns in this case

AThousandShips reviewed (commented):
Will try out this code later today or tomorrow hopefully

AThousandShips reviewed (approved):
Tested briefly and works as expected in my basic testing, except for the earlier code suggestions above this looks good to me but TIWAGOS

One minor usability detail I notice that might be worth to improve is that when you have finished adding a polygon it clears the current tool, so you have to press the add point button again, this was in place originally I'd assume because you replaced the polygon if finishing a new one, but I think you should keep the add point tool now, but that's all the usability details I can speak for in my testing

AThousandShips commented on scene/resources/2d/tile_set.cpp:
```diff
@@ -6710,16 +6798,36 @@ bool TileData::_get(const StringName &p_name, Variant &r_ret) const {
+					}
+					r_ret = get_occluder_polygon(layer_index, 0);
+					return true;
+				} else if (components[1] == "polygons_count") {
```
There's only one other case I can find so consistency is debatable IMO, but you pick, but I'd say it's more easy to find if it's grammatical

AThousandShips replied:
Then true, just saw one here, then we'll leave it ungrammatical for now for consistency with other classes

groud replied:
yeah, too bad we didn't spot those before the move to TileMapLayer. But for now I believe we should stick with the wrongness. :/

groud commented:
Alright, besides the wrong plural. I implemented most suggestions. 
I also added the missing default arguments as @Mickeon reminded me to.
I think it's good to merge.

akien-mga commented:
Thanks!
